File: matatu-sim-python/member3_obstacles_physics.py


import random
import pygame

from member1_setup_config import (
    BASE_H, ROAD_X, ROAD_W, LANE_COUNT, C,
    BASE_SPEED, BRAKE_DECEL, RECOVERY_ACCEL, SPAWN_INTERVAL,
    DETECTION_DISTANCE, CROSSING_CLEAR_SPEED, CROSSING_CLEAR_DISTANCE,
)
from member2_matatu_controls import state, keys_down, lane_center, matatu_rect

VEHICLE_COLORS = ["#d53838", "#de4d3e", "#bc2f4b", "#e65b2d"]


def _c(hex_str):
    hex_str = hex_str.lstrip("#")
    return tuple(int(hex_str[i:i + 2], 16) for i in (0, 2, 4))


def spawn_obstacle():
    if random.random() < 0.52:
        lane = random.randint(0, LANE_COUNT - 1)
        w, h = 82, 62
        state.obstacles.append({
            "kind": "VEHICLE",
            "lane": lane,
            "x": lane_center(lane) - w / 2,
            "y": -h - 20,
            "w": w,
            "h": h,
            "color": _c(random.choice(VEHICLE_COLORS)),
            "detected": False,
            "passed": False,
            "cleared": False,
        })
    else:
        state.obstacles.append({
            "kind": "CROSSING",
            "lane": None,
            "x": ROAD_X + 18,
            "y": -94,
            "w": ROAD_W - 36,
            "h": 74,
            "color": C.white,
            "detected": False,
            "passed": False,
            "cleared": False,
        })


def update_simulation(dt):
    if state.game_over:
        return

    state.braking = "space" in keys_down
    state.elapsed += dt

    if state.braking:
        state.speed = max(0, state.speed - BRAKE_DECEL * dt)
    else:
        state.speed = min(BASE_SPEED, state.speed + RECOVERY_ACCEL * dt)

    state.x += (state.target_x - state.x) * min(1, dt * 12)
    state.distance += state.speed * dt / 12
    state.road_offset = (state.road_offset + state.speed * dt) % 80

    state.spawn_timer += dt
    if state.spawn_timer >= SPAWN_INTERVAL:
        state.spawn_timer = 0
        spawn_obstacle()

    for obstacle in state.obstacles:
        obstacle["y"] += state.speed * dt
        update_detection(obstacle)

        if not obstacle["passed"] and obstacle["y"] > BASE_H + 10:
            obstacle["passed"] = True
            state.avoided += 1

    state.alert_flash = max(0, state.alert_flash - dt)
    state.obstacles = [o for o in state.obstacles if o["y"] < BASE_H + 160]
    check_collision()


def update_detection(obstacle):
    from member1_setup_config import MATATU_Y
    distance = MATATU_Y - (obstacle["y"] + obstacle["h"])

    if 0 <= distance <= DETECTION_DISTANCE:
        obstacle["detected"] = True
        state.alert_flash = 0.2

    if (
        obstacle["kind"] == "CROSSING"
        and obstacle["detected"]
        and not obstacle["cleared"]
        and -obstacle["h"] <= distance <= CROSSING_CLEAR_DISTANCE
        and state.speed <= CROSSING_CLEAR_SPEED
    ):
        obstacle["cleared"] = True


def _rect_of(obstacle):
    return pygame.Rect(int(obstacle["x"]), int(obstacle["y"]), int(obstacle["w"]), int(obstacle["h"]))


def check_collision():
    matatu = matatu_rect()

    for obstacle in state.obstacles:
        if not matatu.colliderect(_rect_of(obstacle)):
            continue

        if obstacle["kind"] == "CROSSING":
            if obstacle["cleared"]:
                continue
            if state.speed <= CROSSING_CLEAR_SPEED:
                obstacle["cleared"] = True
                continue

        state.game_over = True
        state.collision = obstacle
        state.braking = False
        return


def detected_obstacles():
    from member1_setup_config import MATATU_Y
    result = [
        o for o in state.obstacles
        if o["detected"] and o["y"] + o["h"] < MATATU_Y + 140
    ]
    result.sort(key=lambda o: abs(MATATU_Y - (o["y"] + o["h"])))
    return result[:5]

File: member3_obstacles_physics.cpp

#include "member3_obstacles_physics.h"
#include <algorithm>
#include <cstdlib>
#include <cmath>

static sf::Color randomVehicleColor() {
    static const sf::Color colors[] = {
        C::fromHex(0xd53838), C::fromHex(0xde4d3e), C::fromHex(0xbc2f4b), C::fromHex(0xe65b2d)
    };
    return colors[std::rand() % 4];
}

static float randf() {
    return static_cast<float>(std::rand()) / static_cast<float>(RAND_MAX);
}

void spawnObstacle() {
    if (randf() < 0.52f) {
        int lane = std::rand() % LANE_COUNT;
        float w = 82.f, h = 62.f;

        Obstacle obstacle;
        obstacle.kind = "VEHICLE";
        obstacle.lane = lane;
        obstacle.x = laneCenter(lane) - w / 2.f;
        obstacle.y = -h - 20.f;
        obstacle.w = w;
        obstacle.h = h;
        obstacle.color = randomVehicleColor();
        state.obstacles.push_back(obstacle);
    } else {
        Obstacle obstacle;
        obstacle.kind = "CROSSING";
        obstacle.lane = -1;
        obstacle.x = ROAD_X + 18.f;
        obstacle.y = -94.f;
        obstacle.w = ROAD_W - 36.f;
        obstacle.h = 74.f;
        obstacle.color = C::white;
        state.obstacles.push_back(obstacle);
    }
}

void updateDetection(Obstacle& obstacle) {
    float distance = MATATU_Y - (obstacle.y + obstacle.h);

    if (distance >= 0.f && distance <= DETECTION_DISTANCE) {
        obstacle.detected = true;
        state.alertFlash = 0.2f;
    }

    if (obstacle.kind == "CROSSING" &&
        obstacle.detected &&
        !obstacle.cleared &&
        distance >= -obstacle.h &&
        distance <= CROSSING_CLEAR_DISTANCE &&
        state.speed <= CROSSING_CLEAR_SPEED) {
        obstacle.cleared = true;
    }
}

static bool intersects(const sf::FloatRect& a, const sf::FloatRect& b) {
    return a.intersects(b);
}

void checkCollision() {
    sf::FloatRect matatu = matatuRect();

    for (auto& obstacle : state.obstacles) {
        sf::FloatRect obstacleRect(obstacle.x, obstacle.y, obstacle.w, obstacle.h);
        if (!intersects(matatu, obstacleRect)) continue;

        if (obstacle.kind == "CROSSING") {
            if (obstacle.cleared) continue;
            if (state.speed <= CROSSING_CLEAR_SPEED) {
                obstacle.cleared = true;
                continue;
            }
        }

        state.gameOver = true;
        state.collision = Collision{ obstacle.kind };
        state.braking = false;
        return;
    }
}

void updateSimulation(float dt) {
    if (state.gameOver) return;

    state.braking = keysDown.count("space") > 0;
    state.elapsed += dt;

    if (state.braking) {
        state.speed = std::max(0.f, state.speed - BRAKE_DECEL * dt);
    } else {
        state.speed = std::min(BASE_SPEED, state.speed + RECOVERY_ACCEL * dt);
    }

    state.x += (state.targetX - state.x) * std::min(1.f, dt * 12.f);
    state.distance += state.speed * dt / 12.f;
    state.roadOffset = std::fmod(state.roadOffset + state.speed * dt, 80.f);

    state.spawnTimer += dt;
    if (state.spawnTimer >= SPAWN_INTERVAL) {
        state.spawnTimer = 0.f;
        spawnObstacle();
    }

    for (auto& obstacle : state.obstacles) {
        obstacle.y += state.speed * dt;
        updateDetection(obstacle);

        if (!obstacle.passed && obstacle.y > BASE_H + 10.f) {
            obstacle.passed = true;
            state.avoided += 1;
        }
    }

    state.alertFlash = std::max(0.f, state.alertFlash - dt);

    state.obstacles.erase(
        std::remove_if(state.obstacles.begin(), state.obstacles.end(),
                        [](const Obstacle& o) { return o.y >= BASE_H + 160.f; }),
        state.obstacles.end());

    checkCollision();
}

std::vector<Obstacle*> detectedObstacles() {
    std::vector<Obstacle*> result;
    for (auto& obstacle : state.obstacles) {
        if (obstacle.detected && obstacle.y + obstacle.h < MATATU_Y + 140.f) {
            result.push_back(&obstacle);
        }
    }
    std::sort(result.begin(), result.end(), [](Obstacle* a, Obstacle* b) {
        return std::abs(MATATU_Y - (a->y + a->h)) < std::abs(MATATU_Y - (b->y + b->h));
    });
    if (result.size() > 5) result.resize(5);
    return result;
}
